lib/src/rules/parameter_assignments.dart

```diff
@@ -88,9 +136,33 @@ class _Visitor extends SimpleAstVisitor {
       // Getters don't have parameters.
       parameterList.parameters.forEach((e) {
         if (node.body.isPotentiallyMutatedInScope(e.element)) {
-          rule.reportLint(e);
+          _reportIfSimpleParameterOrWithDefaultValue(e, node);
         }
       });
     }
   }
+
+  void _reportIfSimpleParameterOrWithDefaultValue(
+      FormalParameter parameter, AstNode functionOrMethodDeclaration) {
+    final nodes =
+        DartTypeUtilities.traverseNodesInDFS(functionOrMethodDeclaration);
+
+    final assignmentsNodes = nodes.where((n) =>
+        n is AssignmentExpression &&
+        _isDefaultFormalParameterWithoutDefaultValueMutated(parameter, n));
+
+    final nonNullCoalescingAssignments = nodes.where((n) =>
+        n is AssignmentExpression &&
+        _isDefaultFormalParameterWithoutDefaultValueIncorrectlyMutated(
+            parameter, n));
+
+    final shouldReport = parameter is SimpleFormalParameter ||
+        _isDefaultFormalParameterWithDefaultValue(parameter) ||
+        nonNullCoalescingAssignments.isNotEmpty ||
+        assignmentsNodes.length > 1;
+
+    if (shouldReport) {
+      rule.reportLint(parameter);
```

Comment:
The lint should be reported at the site of the assignment, not on the parameter. Otherwise you're asking the user to try to figure out where the assignment is occurring when you could have told them.